source/common/upstream/outlier_detection_impl.cc

```diff
@@ -57,6 +57,12 @@ void DetectorHostSinkImpl::putHttpResponseCode(uint64_t response_code) {
     if (++consecutive_5xx_ ==
         detector->runtime().snapshot().getInteger("outlier_detection.consecutive_5xx",
                                                   detector->config().consecutive5xx())) {
+      // There are two outcomes here. The ejection will be enforced,
+      // or it won't. Either way the host won't trigger the check above if it keeps been
+      // charged with only 5xx ResponseCodes without an intervening non-5xx code. Therefore,
+      // the consecutive_5xx_ counter should be reset to allow the Sink to detect another bout
+      // of 5xx from this host.
+      consecutive_5xx_ = 0;
```

Comment:
It's a minor thing, but to avoid multiple threads thrashing I would do this in onConsecutive5xxWorker